hr_attendance_hour_type_report/models/hr_attendance.py

```diff
@@ -0,0 +1,117 @@
+
+    @api.depends("date")
+    def _compute_date_type(self):
+        for rec in self:
+            if rec.date.weekday() == 6:
+                rec.date_type = "sunday"
+            elif self.env["hr.holidays.public"].is_public_holiday(
+                rec.date, rec.employee_id.id
+            ):
+                rec.date_type = "holiday"
+            else:
+                rec.date_type = "normal"
+
+    @api.depends("check_in")
+    def _compute_date(self):
+        UTC = pytz.timezone("utc")
+        for rec in self:
+            tz = rec.employee_id.tz
+            check_in = rec.check_in
+            if not check_in.tzinfo:
+                check_in = UTC.localize(check_in)
+            check_in_tz = check_in.astimezone(pytz.timezone(tz))
+            rec.date = check_in_tz.date()
+
+    @api.depends("worked_hours", "check_in", "check_out")
+    def _compute_worked_hours_night(self):
+        UTC = pytz.timezone("utc")
+        for rec in self:
+            rec.worked_hours_nighttime = 0
+            rec.worked_hours_daytime = 0
+            if not rec.check_out:
+                continue
+            if rec.worked_hours > 24:
+                raise exceptions.UserError(
+                    _("More than 24h of work in 1 shift is forbidden")
+                )
+            night_start = rec.employee_id.company_id.hr_night_work_hour_start
+            hour_night_start = int(night_start)
+            minute_night_start = int(60 * (night_start - hour_night_start))
+            night_end = rec.employee_id.company_id.hr_night_work_hour_end
+            hour_night_end = int(night_end)
+            minute_night_end = int(60 * (night_end - hour_night_end))
+            tz = pytz.timezone(rec.employee_id.tz)
+            check_in = UTC.localize(rec.check_in)
+            check_out = UTC.localize(rec.check_out)
+            curr_day_night_start = tz.localize(
+                dt.datetime.combine(
+                    rec.date, dt.time(hour=hour_night_start, minute=minute_night_start)
+                )
+            ).astimezone(UTC)
+            curr_day_night_end = tz.localize(
+                dt.datetime.combine(
+                    rec.date, dt.time(hour=hour_night_end, minute=minute_night_end)
+                )
+            ).astimezone(UTC)
+            next_day_night_start = tz.localize(
+                dt.datetime.combine(
+                    rec.date + dt.timedelta(days=1),
+                    dt.time(hour=hour_night_start, minute=minute_night_start),
+                )
+            ).astimezone(UTC)
+            next_day_night_end = tz.localize(
+                dt.datetime.combine(
+                    rec.date + dt.timedelta(days=1),
+                    dt.time(hour=hour_night_end, minute=minute_night_end),
+                )
+            ).astimezone(UTC)
+            # import pdb; pdb.set_trace()
+            rec.worked_hours_nighttime += (
+                min(check_out, curr_day_night_end) - min(curr_day_night_end, check_in)
```

Comment:
```suggestion
                min(check_out, curr_day_night_end) - min(curr_day_night_start, check_in)
```

i am not sure about this?

Reply:
forgetting about TZ for a moment

Night starts at 22:00 and ends at 6:00. 

```
min(check_out, curr_day_night_end) - min(curr_day_night_end, check_in)
```
checkin at 3am, checkout at 12:00 -> min(check_out=12:00, curr_day_night_end=6:00) = 6:00 and min(curr_day_night_end, check_in) = 3:00 -> we have 3h of night time on that slot

The computation for a checking at 17:00 , checkout at 23:00 happens in the if block just below.